content/blog/2019-12-18-extensible/index.md

```diff
@@ -0,0 +1,250 @@
+
+The generics for the instruction and function types were also added to increase type safety within the main evaluation loop and within the handler functions. Having the `pc` parameterized on those types allowed me to write code that knew that each function would have a list of instruction and a name. However, because the Bril program is parsed from a JSON file and then cast to a Bril program type without any dynamic checks means that this type safety was largely negated by the initial cast. More fundamentally, the way most extensions check that the instruction passed into `evalInstrs` is pretty unsound. The `instr` type that `evalInstrs` usually takes in is given the following type:
+
+```javascript
+instr: Instruction | I
+```
+
+where `I` is the generic type and `Instruction` is the type of the instructions that the extension implements. The `isInstruction` function is usually just a function that returns a type guard if the opcode of the instruction matches one of the implemented instruction ops (here in the `instrOp` array):
+
+```javascript
+function isInstruction(instr: BaseInstruction): instr is Instruction {
+  return instrOps.some(op => op === instr.op);
+}
+```
+
+This is not really a safe type guard because one of your instructions could assume a `dest` field in the instruction but this is only checking that the opcode matches, not that the instruction has all of the correct fields. This could maybe be improved by more comprehensive runtime type checking, but would still probably not rule out everything that could go wrong. Furthermore, it could do the wrong thing if extensions are composed in the wrong way. If two extensions `A` and `B` implement operation `a` but an extension `A` removes some fields of operation `a`, if `B` extends `A`, then `B` will be operating on the wrong kind of instruction. However, if `A` extends `B`, then the `a` operation will be handled by extension `A` and never be propagated to `B`.
+
+Furthermore, because typescript doesn't really support transforming discriminated unions of strings into string arrays, the `instrOps` array above unfortunately can't be set to something like `Instruction["op"]`. This could lead to other issues if some item in `instrOps` is typed in incorrectly, or if the extension is modified to support more or less instructions (the developer would also have to remember to update this array). One bug that did happen is that I forgot to include the `"const"` field in the base extensions `instrOps` array. In order to slightly remedy this, I came up with the following code snippet to make sure string arrays are equivalent to discriminated unions of strings:
```

Comment:
Hmm, I’m a little confused here... I’m not sue where the “discriminated unions of strings” are here, and I don’t exactly know what the type `Instruction["op"]` would mean here.